src/frontend/src/script.rs

```diff
@@ -78,30 +84,76 @@ mod python {
     use crate::error::{CatalogSnafu, InvalidSystemTableDefSnafu, TableNotFoundSnafu};
     use crate::instance::Instance;
 
+    /// A placeholder for the real gRPC handler.
+    /// It is temporary and will be replaced soon.
+    struct DummyHandler;
+
+    impl DummyHandler {
+        fn arc() -> Arc<Self> {
+            Arc::new(Self {})
+        }
+    }
+
+    #[async_trait::async_trait]
+    impl GrpcQueryHandler for DummyHandler {
+        type Error = Error;
+
+        async fn do_query(
+            &self,
+            _query: Request,
+            _ctx: QueryContextRef,
+        ) -> std::result::Result<Output, Self::Error> {
+            unreachable!();
+        }
+    }
+
     pub struct ScriptExecutor {
-        script_manager: ScriptManager,
+        script_manager: ScriptManager<Error>,
+        grpc_handler: ArcSwap<FrontendGrpcQueryHandlerRef>,
+        catalog_manager: CatalogManagerRef,
     }
 
     impl ScriptExecutor {
         pub async fn new(
             catalog_manager: CatalogManagerRef,
             query_engine: QueryEngineRef,
         ) -> Result<Self> {
+            let grpc_handler = DummyHandler::arc();
             Ok(Self {
-                script_manager: ScriptManager::new(catalog_manager, query_engine)
+                grpc_handler: ArcSwap::new(Arc::new(grpc_handler.clone() as _)),
+                script_manager: ScriptManager::new(grpc_handler as _, query_engine)
                     .await
                     .context(crate::error::StartScriptManagerSnafu)?,
+                catalog_manager,
             })
         }
 
-        pub async fn start(&self, instance: &Instance) -> Result<()> {
+        pub fn start(&self, instance: &Instance) -> Result<()> {
+            let handler = Arc::new(instance.clone());
+            self.grpc_handler.store(Arc::new(handler.clone() as _));
+            self.script_manager
+                .start(handler)
+                .context(crate::error::StartScriptManagerSnafu)?;
+
+            Ok(())
+        }
+
+        /// Create scripts table for the specific catalog if it's not exists.
+        /// The function is idempotent and safe to be called more than once for the same catalog
+        async fn create_scripts_table_if_need(&self, catalog: &str) -> Result<()> {
+            let scripts_table = self.script_manager.get_scripts_table(catalog);
+
+            if scripts_table.is_some() {
+                return Ok(());
+            }
+
             let RegisterSystemTableRequest {
                 create_table_request: request,
                 open_hook,
-            } = self.script_manager.create_table_request();
+            } = self.script_manager.create_table_request(catalog);
 
-            if let Some(table) = instance
-                .catalog_manager()
+            if let Some(table) = self
```

Comment:
Maybe need to acquire write lock first.

Reply:
`create_scripts_table_if_need` is idempotent, calling it many times doesn't matter. The only side effect is that creates some unused instances.